### PR TITLE
stretch-goals

### PR DESCRIPTION
**Pull request by haroldhyte**
added user roles and restricted page views and actions accordingly
added company-specific views for billing records and invoices
fixed date sorting
various stylistic changes to simplify the ui and add consistency
added table sorting to invoices

### CODE REVIEW

File: src/app/billing-record/billing-record.component.ts [hunk 1]
```diff
@@ -53,6 +55,19 @@ sortBy(category){
       if( category === 'createdBy'){
         return this.billingRecords.sort((a,b)=> a.createdBy.username.localeCompare(b.createdBy.username))
       }
+      if( category === 'dueDate'){
+        return this.billingRecords.sort((a,b) =>
+          {
+            let left = new Date(a.dueDate).getUTCDate();
+            let right = new Date(b.dueDate).getUTCDate();
+            if(left < right) {
+              return -1;
```
Comment (jawitzke) [inline src/app/billing-record/billing-record.component.ts]: what are these numbers

File: src/app/billing-record/billing-record.component.ts [hunk 2]
```diff
@@ -69,10 +84,32 @@ sortBy(category){
...
         return this.billingRecords.sort((a,b)=> b.createdBy.username.localeCompare(a.createdBy.username))
       }
+      if( category === 'dueDate'){
+        return this.billingRecords.sort((a,b) =>
+          {
+            let left = new Date(a.dueDate).getUTCDate();
+            let right = new Date(b.dueDate).getUTCDate();
+            if(left > right) {
+              return -1;
+            } else if (left === right) {
```
Comment (jawitzke) [inline src/app/billing-record/billing-record.component.ts]: can all of this repeat logic be moved to a common function?

File: src/app/billing-record/company-billing-record.component.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        if( category === 'createdBy'){
+          return this.billingRecords.sort((a,b)=> a.createdBy.username.localeCompare(b.createdBy.username))
+        }
+        if( category === 'dueDate'){
+          return this.billingRecords.sort((a,b) =>
+            {
+              let left = new Date(a.dueDate).getUTCDate();
+              let right = new Date(b.dueDate).getUTCDate();
+              if(left < right) {
+                return -1;
```
Comment (jawitzke) [inline src/app/billing-record/company-billing-record.component.ts]: more of that same repeat logic

File: src/app/billing-record/company-billing-record.component.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        if( category === 'createdBy'){
+          return this.billingRecords.sort((a,b)=> b.createdBy.username.localeCompare(a.createdBy.username))
+        }
+        if( category === 'dueDate'){
+          return this.billingRecords.sort((a,b) =>
+            {
+              let left = new Date(a.dueDate).getUTCDate();
+              let right = new Date(b.dueDate).getUTCDate();
+              if(left > right) {
+                return -1;
```
Comment (jawitzke) [inline src/app/billing-record/company-billing-record.component.ts]: again

File: src/app/invoice/company-invoice.component.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+          return this.invoices.sort((a,b)=> a.createdBy.username.localeCompare(b.createdBy.username))
+        }
+        if( category === 'dueDate'){
+          return this.invoices.sort((a,b) =>
+            {
+              let left = new Date(a.dueDate).getUTCDate();
+              let right = new Date(b.dueDate).getUTCDate();
+
+              if(left < right) {
+                return -1;
```
Comment (jawitzke) [inline src/app/invoice/company-invoice.component.ts]: for real please a common parent function

File: src/app/invoice/company-invoice.component.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            }
+          );
+        }
+        if( category === 'createdOn'){
+          return this.invoices.sort((a,b) =>
+            {
+              let left = new Date(a.createdOn).getUTCDate();
+              let right = new Date(b.createdOn).getUTCDate();
+              if(left < right) {
+                return -1;
```
Comment (jawitzke) [inline src/app/invoice/company-invoice.component.ts]: again

File: src/app/invoice/company-invoice.component.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        if( category === 'createdBy'){
+          return this.invoices.sort((a,b)=> b.createdBy.username.localeCompare(a.createdBy.username))
+        }
+        if( category === 'dueDate'){
+          return this.invoices.sort((a,b) =>
+            {
+              let left = new Date(a.dueDate).getUTCDate();
+              let right = new Date(b.dueDate).getUTCDate();
+              if(left > right) {
+                return -1;
```
Comment (jawitzke) [inline src/app/invoice/company-invoice.component.ts]: one more time

File: src/app/invoice/company-invoice.component.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+              } else return 1;
+            }
+          );
+        }
+        if( category === 'createdOn'){
+          return this.invoices.sort((a,b) =>
+            {
+              let left = new Date(a.createdOn).getUTCDate();
+              let right = new Date(b.createdOn).getUTCDate();
+              if(left > right) {
```
Comment (jawitzke) [inline src/app/invoice/company-invoice.component.ts]: one more time

File: src/app/billing-record/billing-record.component.ts (hunk 1 above)
Comment (haroldhyte) [reply src/app/billing-record/billing-record.component.ts]: this is the standard implementation for a comparison function in Javascript

File: src/app/billing-record/billing-record.component.ts (hunk 2 above)
Comment (haroldhyte) [reply src/app/billing-record/billing-record.component.ts]: yes, will do

**Pull request merged**